### ISSUE TITLE
Linear History

### ISSUE THREAD

**Issue by kousu** [shimming-toolbox/shimming-toolbox#14]
I turned on ["Linear History" for `master`](https://github.com/shimming-toolbox/shimming-toolbox-py/settings/branch_protection_rules/16666687). If it becomes too much of a headache we can turn it off, but I think it will pay off in the long run when we need to go code-spelunking in the future to get in the habit now.

This means:

* when working on a pull request, always do `git fetch; git rebase origin/master` to sync up (as with `git merge master` if there are conflicts you will need to solve them).
* everyone should `git config pull.rebase true` so that when working on a branch with someone else, use `git pull` makes a linear history.

**Comment by mathieuboudreau** [comment]
@kousu does `git config pull.rebase true` change your global git config, or just for the current git repo. All the other projects I work on use squash merges and not rebases (rebase is rarely recommended for collaborative projects since it rewrites the git history), so I don't want to mess up my git history for all the other repos I work on by changing my global config preferences.

**Comment by jcohenadad** [comment]
@mathieuboudreau i was wondering the same thing-- when running the command outside of a repos, i get an error, so i concluded that this command only affects the repos where the command is run (but i could be wrong...)

**Comment by jcohenadad** [comment]
@kousu given that our team (i include myself) is not expert in git, the past couple of months/years have demonstrated that resolving conflicts when using rebase can take up time that could otherwise be dedicated to advancing science, e.g. [here](https://github.com/shimming-toolbox/shimming-toolbox-py/pull/5#issuecomment-656235344). 

i would thus be in favour of switching back to the "git merge" approach, which the team is more accustomed to.

**Comment by gab-berest** [comment]
I'm not entirely sure, but I think it only rewrites in the current branch (in your .git/config file). Using --global would make a rewrite in the /.gitconfig that is global to all repository (you don't want that).
I'm alos quit accustomed to the merge style rather than the rebase one. Normally, if everybody respects good git commands and is carefull, it should be easy to work with.

**Comment by kousu** [comment]
I'm going to advocate for this and if you're still unsure afterwards I'm happy to turn it off (or show you where to: it's just in https://github.com/shimming-toolbox/shimming-toolbox-py/settings/branch_protection_rules/16666687; I think everyone has been given the rights to change that already).

I think if you give it a shot you'll find linear history keeps the project manageable. Like https://medium.com/singlestone/a-git-workflow-using-rebase-1b1210de83e5 says 

> Git can be tricky to use, especially in a team setting where stale branches and merge conflicts tend to cause problems when you least have time for them. I’ve seen many teams turn their repository into a tangled mess as they struggle with Git. The result is always a lot of wasted time, an indecipherable log of what actually happened in the repository, and very little confidence that the resulting codebase contains the changes it should. [...]
> [Rebase is] a way to cleanly incorporate new code into a feature branch and a way to keep meaningless commit history out of a repository’s master branch. The end result is a linear commit history on the master branch, which makes it easier to see how the code evolved.

That is, a linear history will pay off in debugging and code archaeology in the future.

For an example of how merging can go bad, https://github.com/neuropoly/spinalcordtoolbox/pull/2751/commits got lost in our inboxes, so a 1 line change ended up needing to be resync'd with `master` 5 times, each of those showing up as a commit in the `git log` and adding extra lines in the visualization you get in `git log --graph`. Rebasing preserved the original intent.

Another way it can go bad is that fixes made during a merge conflict (which are potentially very extensive) get hidden by the git UI because merge commits are a special case. It's not obvious where certain code even came from sometimes if someone felt they had to make extensive fixes to get a merge conflict to go away.

---

You've used

```
git checkout master
git pull
git checkout feature-branch
git merge master
# fix merge conflicts
git push
```

The only difference is:

```
git checkout master
git pull
git checkout feature-branch
git rebase master
# fix merge conflicts
git push --force
```

> @kousu given that our team (i include myself) is not expert in git, the past couple of months/years have demonstrated that resolving conflicts when using rebase can take up time that could otherwise be dedicated to advancing science

@jcohenadad I know we had some trouble with rebasing in the past. I think you are misattributing the problem. Merge conflicts are going to happen whenever two branches diverge. With merging, you have to fix the conflicts all at once in a single commit. On the other hand, rebasing makes you fix the merge conflicts commit-by-commit, which can take longer, though that is offset by conflict being smaller. The problem we had the day I mostly failed at trying to help you with your branch was that some of your commits had been pulled out to separate PRs.

So merging seems like less work in the short term, because you only need to do one batch of conflicts, but with rebasing it's less likely you'll introduce bugs when you fix the conflicts, and importantly you preserve **the meaning** of each commit for future readers.

Merging is meant for a truly decentralized project where everyone has their own version customized to their liking but they want to be able to benefit from each others' work. Here, we have a shared repo where everyone has write access to it, so we can choose to be disciplined about keeping history linear.

--

> (rebase is rarely recommended for collaborative projects since it rewrites the git history)

It's only dangerous to do this for shared branches like `master`, and that's why https://github.com/shimming-toolbox/shimming-toolbox-py/settings/branch_protection_rules/16666687 protects against that too. It is totally fine to rewrite the history for a feature branch; as you go through code review I even think you *should* be rewriting its history -- changing commit messages, squashing (`git rebase -i` => "squash") and splitting commits (`git rebase -i` => "edit" => `git reset -p` to pick out which parts of a commit should be separated out to become a new commit), to clean up mistakes and detours and make it more obvious to future readers what your overall intent turned out to be, once you've incorporated all the feedback from code review.

---

> @kousu does `git config pull.rebase true` change your global git config, or just for the current git repo. All the other projects I work on use squash merges and not rebases, so I don't want to mess up my git history for all the other repos I work on by changing my global config preferences.

It will edit `.git/config` which is just for your local copy of the project. Running `git config --global ....` will edit `~/.gitconfig` which would affect all your projects. But this is just a shortcut for `git pull --rebase`, and you don't even need to use that unless you and someone else are working on a branch together.

---

By the way, a squash is a kind of rebase. If you run `git rebase -i <something>` in a project you'll get this help:

```
# p, pick = use commit
# r, reword = use commit, but edit the commit message
# e, edit = use commit, but stop for amending
# s, squash = use commit, but meld into previous commit
# f, fixup = like "squash", but discard this commit's log message
# x, exec = run command (the rest of the line) using shell
# d, drop = remove commit
```

When you click the "squash merge" button on github what's happening it's writing a rebase list full of `s` commands. I don't really think Github expects squash merging to be a Best Practice; I expect it's meant for accepting small, kind of messy PRs from people outside a project full of lots of "oops" and "sorry!" and "Updated README.md" commits; or for removing first-draft code that was eventually changed during code review. The default merge button tries to keep history linear; the only time it can't is if your PR branch has lasted long enough that you've had to resync with `master` one or a few times and you've done it with `git merge master`.

**Comment by jcohenadad** [comment]
> @jcohenadad I know we had some trouble with rebasing in the past. I think you are misattributing the problem. Merge conflicts are going to happen whenever two branches diverge. 

indeed, i wasn't clear, thank you for clarifying.

> With merging, you have to fix the conflicts all at once in a single commit. On the other hand, rebasing makes you fix the merge conflicts commit-by-commit, which can take longer, though that is offset by conflict being smaller. 

right, i think this is precisely my argument for advocating the merge. I usually never look at previous commits, only the latest version, and fix conflicts by comparing version X and version Y and manually picking what i want. As you said, we only need to do it once with 'merge'. 
With rebasing, if the commit sequence is messy (added things there, then removing, then added again, etc.), we end us spending 1+ hour fixing every commit by doing the above-mentioned task several time, whereas a merge-conflict usually takes 5 minutes. If we multiply 1h by the number of merge/rebase that are done in a project's lifetime, that's a lot. 

So, my understanding (correct me if i'm wrong), is that the decision for merge/rebase is a balance between these two arguments:
- rebase: we have a clean commit history
- merge: we only need to fix commits once, the procedure is faster, especially for people unfamiliar with git.

I do appreciate your perspective of doing things the right way, but at the same time I've been witnessing all sorts of problems related to git usage in the lab. I understand that people can (and should) be educated with git, but the learning curve is slow, and an academic lab often has short-term trainees (often not proficient with git) who need to focus most of their attention on the underlying science and software implementation. Adding this level of complexity is, I think, too much to ask and will discourage our valuable contributors (e.g. [here](https://github.com/shimming-toolbox/shimming-toolbox-py/pull/5#issuecomment-656235344)).

**Comment by kousu** [comment]
I've fixed the conflict in #5 and here's how I did it:


1. Sync with Github

```
$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
$ git pull
Enter passphrase for key '/home/kousu/.ssh/id_ed25519.gitlab': 
remote: Enumerating objects: 92, done.
remote: Counting objects: 100% (92/92), done.
remote: Compressing objects: 100% (48/48), done.
remote: Total 84 (delta 29), reused 79 (delta 28), pack-reused 0
Unpacking objects: 100% (84/84), done.
From github.com:shimming-toolbox/shimming-toolbox-py
   61ddcfc..a317a3b  master          -> origin/master
 * [new branch]      ad/unwrap-phase -> origin/ad/unwrap-phase
 * [new branch]      gc/dcm2bids     -> origin/gc/dcm2bids
Updating 61ddcfc..a317a3b
Fast-forward
 .gitignore       |  1 +
 .travis.yml      |  2 +-
 README.md        | 27 +++++++++++++++++++++++++++
 requirements.txt |  3 ---
 setup.py         | 15 +++++++--------
 5 files changed, 36 insertions(+), 12 deletions(-)
 delete mode 100644 requirements.txt
Current branch master is up to date.
```

2. Rebase

```
$ git checkout mb/sim
Branch 'mb/sim' set up to track remote branch 'mb/sim' from 'origin'.
Switched to a new branch 'mb/sim'
$ git rebase master
First, rewinding head to replay your work on top of it...
Applying: Create empty class
Applying: Add default zeros volume
Applying: Add basic shepp-logan init & test
Using index info to reconstruct a base tree...
A	requirements.txt
Falling back to patching base and 3-way merge...
CONFLICT (modify/delete): requirements.txt deleted in HEAD and modified in Add basic shepp-logan init & test. Version Add basic shepp-logan init & test of requirements.txt left in tree.
error: Failed to merge in the changes.
Patch failed at 0003 Add basic shepp-logan init & test
Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```

3. Investigate the merge conflicts

```
$ git diff
* Unmerged path requirements.txt
$ git am --show-current-patch
commit 71404abedadca7c27ae76a4af6bf16911e6b79c8
Author: Mathieu Boudreau <emb6150@gmail.com>
Date:   Fri Jul 3 22:10:59 2020 -0300

    Add basic shepp-logan init & test

diff --git a/requirements.txt b/requirements.txt
index a6d4379..c32bd04 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -1,2 +1,4 @@
 pytest==4.6.3
 pytest-cov==2.5.1
+numpy==1.19.0
+phantominator==0.6.4

[...clipped...]
```

Now I know what happened here: in #7 I moved the contents of requirements.txt into setup.py directly, and changed the recommendation to `pip install -e .` instead of `pip install -r requirements.txt` (because that makes for a more consistent python env). So now I know that I need to take your patch to requirements.txt and translate it into the right place in setup.py; and also that the way to resolve the edit/delete requirements.txt conflict is to choose delete:

4. Resolve the conflict

```
$ git rm requirements.txt 
requirements.txt: needs merge
rm 'requirements.txt'
$ vi setup.py # paste in the changes in the right place
$ git diff    # review
diff --git a/setup.py b/setup.py
index e4e5de9..03c6480 100644
--- a/setup.py
+++ b/setup.py
@@ -21,7 +21,8 @@ setup(
     keywords="",
     packages=find_packages(exclude=["contrib", "docs", "tests"]),
     install_requires=[
-        #"numpy~=1.16.0",
+        "numpy~=1.19.0",
+        "phantominator~=0.6.4",
     ],
     extras_require={
         'testing': ["pytest~=4.6.3", "pytest-cov~=2.5.1"]
$ git add -u
$ git rebase --continue   # commit
```

5. Apply the rest of the changes in the branch

```
$ git rebase --continue   # (continued from above)
Applying: Add basic shepp-logan init & test
Applying: Add customized phantom and tests
Applying: Add method to generate deltaB0 & tests
Applying: Add simulation and get methods & tests
Applying: Add SNR functionality & test
Applying: Add sanity check for simulation - B0 map measurement must match generated one
Applying: Add test for right-handedness
Applying: Add save functionality + test
Using index info to reconstruct a base tree...
A	requirements.txt
Falling back to patching base and 3-way merge...
CONFLICT (modify/delete): requirements.txt deleted in HEAD and modified in Add save functionality + test. Version Add save functionality + test of requirements.txt left in tree.
error: Failed to merge in the changes.
Patch failed at 0010 Add save functionality + test
Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```

Ah a second merge conflict. Gotta deal with that too.

This is what I meant about resolving conflicts one by one, but each conflict being smaller.

6. Investigate this conflict

```
$ git diff
* Unmerged path requirements.txt
$ git am --show-current-patch   # this tool was recommended by `git rebase`
commit 8a8b9e3da687da3c503d00549082b8d000047683
Author: Mathieu Boudreau <emb6150@gmail.com>
Date:   Sat Jul 4 02:21:44 2020 -0300

    Add save functionality + test

diff --git a/requirements.txt b/requirements.txt
index c32bd04..ad8369b 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -2,3 +2,5 @@ pytest==4.6.3
 pytest-cov==2.5.1
 numpy==1.19.0
 phantominator==0.6.4
+nibabel==3.1.1
+scipy==1.5.0
[...clipped...]
```

7. Fix the conflict

```
$ git rm requirements.txt 
requirements.txt: needs merge
rm 'requirements.txt'
$ vi setup.py  # paste in the nibabel/scipy line 
$ git diff     # verify the fix
diff --git a/setup.py b/setup.py
index 03c6480..1bf0314 100644
--- a/setup.py
+++ b/setup.py
@@ -23,6 +23,8 @@ setup(
     install_requires=[
         "numpy~=1.19.0",
         "phantominator~=0.6.4",
+        "nibabel~=3.1.1",
+        "scipy~=1.5.0",
     ],
     extras_require={
         'testing': ["pytest~=4.6.3", "pytest-cov~=2.5.1"]
$ git add -u
$ git rebase --continue  # commit
```

8. Keep applying the patches

```
$ git rebase --continue  # (continued from above)
Applying: Add save functionality + test
Applying: Add more tests
Applying: Add general signal tests
Applying: Ran Black code auto-formatter
Applying: Document class & adhere names to convention
Applying: Fix typo
```

Hooray! We're done.

9. Update Github

```
$ git push -f
Enter passphrase for key '/home/kousu/.ssh/id_ed25519.gitlab': 
Counting objects: 114, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (88/88), done.
Writing objects: 100% (114/114), 19.90 KiB | 3.98 MiB/s, done.
Total 114 (delta 39), reused 82 (delta 25)
remote: Resolving deltas: 100% (39/39), completed with 1 local object.
To github.com:shimming-toolbox/shimming-toolbox-py.git
 + e554a3e...4cc5e3b mb/sim -> mb/sim (forced update)
```


The merge workflow would have been very similar, except that I would have done a single merge commit where I pasted all four new dependencies into setup.py and did `git rm requirements.txt` once. That can work too, but that tends to leave a lot of guesswork involved at each merge conflict. The advantage of this way is that now the [first mention](phantominator) of numpy/phantominator in the code is also the first mention in setup.py; this is highlighted if you look at [`git blame`](https://github.com/shimming-toolbox/shimming-toolbox-py/blame/4cc5e3bcabbaae0c13b0ecf127aaf7327a735951/setup.py#L24-L25) (and [same for](https://github.com/shimming-toolbox/shimming-toolbox-py/blame/4cc5e3bcabbaae0c13b0ecf127aaf7327a735951/setup.py#L26-L27) scipy/nibabel), keeping those lines together as a coherent meaningful unit. And [here's that linear history](https://github.com/shimming-toolbox/shimming-toolbox-py/network):

![net](https://user-images.githubusercontent.com/987487/87231676-1c3d4480-c387-11ea-9332-d9efe86d315c.png)

**Comment by kousu** [comment]
> So, my understanding (correct me if i'm wrong), is that the decision for merge/rebase is a balance between these two arguments:
> 
>     * rebase: we have a clean commit history
> 
>     * merge: we only need to fix commits once, the procedure is faster, especially for people unfamiliar with git.
> 
> 

But I would add: 

* merge: you're less sure you've made the right changes, because there are changes no longer tethered to the commit message that introduced them

exactly because

> I usually never look at previous commits, only the latest version, and fix conflicts by comparing version X and version Y and manually picking what i want.

> With rebasing, if the commit sequence is messy (added things there, then removing, then added again, etc.), we end us spending 1+ hour fixing every commit by doing the above-mentioned task several time, whereas a merge-conflict usually takes 5 minutes. If we multiply 1h by the number of merge/rebase that are done in a project's lifetime, that's a lot.

That can -- and should -- be mitigated by syncing with `master` often (e.g. at least once a day), something that also helps the `merge` workflow -- and by getting comfortable with `git rebase -i`'s `reword`, `edit`, `fixup` and commit re-ordering features.

> I do appreciate your perspective of doing things the right way, but at the same time I've been witnessing all sorts of problems related to git usage in the lab. I understand that people can (and should) be educated with git, but the learning curve is slow

That's fair. Git is really not that well documented for something so widely used. I think the Git UX is pretty bad, to be honest ([e.g.](http://people.csail.mit.edu/sperezde/onward13.pdf) or [this](https://news.ycombinator.com/item?id=12785200)). They should have made merge commits the exception rather than the rule, been more stern about pushing the built in [`git send-email`](https://git-send-email.io/) command as the default way to collaborate, and been more proscriptive on just what branches were *for* so that people didn't get into endless confusion and debate over what the best workflow is.

How about we disable the protection rule, but during code review I'll link people back to this thread to suggest they give it a shot?

**Comment by jcohenadad** [comment]
> How about we disable the protection rule, but during code review I'll link people back to this thread to suggest they give it a shot?

👍

**Comment by jcohenadad** [comment]
@kousu sorry but i encountered another issue with rebasing, and no time to look into this-- (https://github.com/shimming-toolbox/shimming-toolbox-py/pulls) so i am now allowing squash-merge

**Comment by kousu** [comment]
@jcohenadad I fixed the merge conflicts in https://github.com/shimming-toolbox/shimming-toolbox-py/tree/ko/dcm2bids. I can merge that to @gaspardcereza's branch if it looks right to you.

**Comment by jcohenadad** [comment]
>  if it looks right to you

i don't have the time to inspect, so please do merge if you think it is fine-- thanks!

**Comment by jidicula** [comment]
Necromancing a thread, but I came across this interesting Medium article about potential pitfalls of rebasing (as well as prioritizing a linear history): https://medium.com/@fredrikmorken/why-you-should-stop-using-git-rebase-5552bee4fed1.

Thoughts?

**Comment by kousu** [comment]
The one pitfall that article highlights is that an abstract conflict, the kind git can't catch, like a deleted/added dependency or a moved header file, will cause a broken part of the log. You'll discover it after the rebase when you try to use the code and then will have to add a fixup commit.

But that will happen with a merge too. You'll need to add a fixup to get things working again. So

> how can we avoid these chains of broken commits during rebasing?

is a red herring. Rebase isn't the cause of the brokenness. They're misattributing their frustrations.

They briefly acknowledge that

> One approach could be to let the rebase process finish, test the code to identify any bugs, and go back in history to fix the bugs where they were introduced. To this end, we could use interactive rebasing.

And yes, this is exactly what I do. I write a fixup commit, and then run `git rebase -i` a second time to move the fixup back in time to where it belongs. It's part of the process of [explaining to your future selves](https://dhwthompson.com/2019/my-favourite-git-commit) to think about how the whole chain looks when you integrate code. Part of the rebase workflow is making sure that the rebases are correct; if their coworker missed that that's as much a bug as making the tests fail, and code review should be looking out for that.

---

A more minor point they raise, not a pitfall exactly, is the claim that

> [fixing conflicts/bugs during rebasing] is a cumbersome and error-prone process [...]. Is there a simpler and better way? There is; Git merge. It’s a simple, one-step process, where all conflicts are resolved in a single commit.

And I think exactly the opposite and this is the main reason I am more comfortable with rebases than not. Yes they are quicker to *commit*, but they are much slower to be confident in. With rebasing you get to fix conflicts step by step in bite sized chunks, whereas merging leaves you guessing and leaning on your knowledge of how it "should look", which isn't always right.

And then in the future, those commits litter the history, each potentially having introduced any amount of arbitrary new code hidden behind an innocuous "Merge master" messages. Someone -- that someone probably being you -- reading back to figure it out will be lost. Having smaller messages like "Upgrade libzazz to get new feature X we're going to use" inserted back in the history highlights why and where things happened.

---

Another minor point they raise is

> The resulting merge commit clearly marks the integration point between our branches, and our history depicts what actually happened, and when it happened.

And having an accurate log is nice in principle, I'll give it that. And it makes sense in a world where multiple forks are living in parallel and trading code back and forth but with none being in charge. But I don't think it makes sense in trunk-based development. For that, our integration points are all: when a pull request goes in. Up to that point code review finds and fixes bugs, including the integration bugs. The history of the feature branches is not important to us, it's the story of `master` that we care about.

It's not lying to leave your draft notes on the side when preparing a paper or a proof for publication, it's being helpful to the reader.


--

> I’ve come to the conclusion that it’s about vanity. Rebasing is a purely aesthetic operation. The apparently clean history appeals to us as developers, but it can’t be justified, from a technical nor functional standpoint.

Programming is all about aesthetics. Tidier code means and easier to manage project, and a tidier history is the same. Good aesthetics make a project that's a joy instead of a grind to work on.

---

Maybe the debate is about front-loading vs back-loading effort. With rebasing, you front-load the effort to keep the project tidy, and you think about crafting your commit messages the same way you craft good comments or technical documentation, meaning you have less clean ups and uncertainty in the future. With merging, you get code in and active faster, and probably in use and tested faster, maybe but it can spiral out of your control and cause a lot of black-box testing to figure out what's going on down the road.

---

But also this is like vi vs emacs or tabs vs spaces so just do whatever feels good to the team. I'll help support people in either workflow.

**Comment by jcohenadad** [comment]
i'm still in favour of @kousu 's arguments for using linear history; I really like the concept. But I also think we need to be realistic/pragmatic. So far it has caused us more time to do it properly, and when not done properly, it causes more harm than it does good. Eg:
- inclusion of binaries in PRs (followed by their removal-- which should be followed by a rebase to make sure they disappear from history-- this "follow up" requires maintenance/supervision: this is human time which could be used elsewhere)
- exploration commits: "trying this, trying that, doesn't work, fix this, fix again that, forgot typo, etc.": i know we should not do that, but we do. The reason we do, is again because of lack of available time-- not laziness. 
- team not being used to rebasing (eg: https://github.com/shimming-toolbox/shimming-toolbox-py/pull/73). This is not the first time, not the last. This is how it is, and i would really prefer if people would spend more time on how to progress on the codebase rather than figuring out rebasing problems.

**Comment by jidicula** [comment]
Ok, I have a bunch of jumbled thoughts about this that might lead us to a good workflow.

(*This is a bit tangential, but I promise it'll be relevant*) For the most part, there is a nonlinear relationship between PR size and the lifespan of the corresponding feature branch. Once you have to spend more than 30 minutes reviewing a large PR, your ability to read the code critically goes down drastically. In contrast, smaller PRs are easier to review and can therefore be integrated faster.

When you cut down on the lead time between PR submission for review and branch integration into trunk, you can reduce the likelihood of having to grab patches from trunk to put into your feature branch, which is done either by merging trunk into feature-branch, or rebasing feature-branch onto master. Both approaches are still susceptible to the same pitfalls, as @kousu said above. We should therefore prioritize small PRs to help keep us out of the weeds. This isn't a silver bullet, but I think encouraging this small behavioural change on the part of contributors can lead to large gains with PR lead time because of that nonlinear aspect.

I have 2 mental categories for PRs in a project:

1. When things go right, we want contributors to write a feature well, make a PR, receive quick and good feedback from reviewers, address change requests, and finally integrate their branch into the trunk. The ease of the last step is key – when things are going right, we want to make sure there are no hurdles making a simple process more complex. So far, it seems like even when things go right, the last step goes wrong because of the fragility of the rebase approach and our team's unfamiliarity with it. We're still losing time and bandwidth for what should be a simple operation.

2. When things go wrong (i.e. large PR -> takes longer to review -> takes longer to address changes -> longer branch lifespan -> branch is out of sync with master), having to grab features from the trunk is unavoidable. This is where the real trouble with merge commits emerges and leads to a messy history, and where a rebasing approach can really shine.

I think our goal should be to streamline (and therefore incentivize) scenario 1 as much as possible (merge commits only from feature branch into master), while also using the correct tool (rebase feature branch onto master) to prevent scenario 2 from getting out of hand with respect to our Git history.

With this hybrid approach, when viewing the log, a reader can quickly see that any merge commit corresponds to the merging of a PR and therefore inclusion of a new feature, which is followed by a series of (hopefully) atomic commits. GitHub titles its merge commits with the PR number and the name of the feature branch, which allows for 1-glance forensics if you're trying to track down when a bug or feature was introduced into the codebase. It also hotlinks to the actual PR discussion, which is helpful for understanding the reasoning behind any additional changes that were added, as well as any CI failures. In our project (and probably most others), I'd say that most of the important discussion leading to a feature or bugfix integration happens in PR threads. Unfortunately, when a project uses the rebase workflow for integrating feature branches into the trunk, the relevant PR isn't immediately visible in the GitHub UI for viewing commits (it's 1 or 2 further clicks away).

Rebasing under this hybrid approach is then only used when a feature branch is too long-lived and a patch in the trunk is needed for the feature branch to build correctly. If this scenario only occurs rarely, the frequency of rebase operations goes down, reducing time spent on trying to untangle a botched rebase. 

I think this can also respect our need/want for a somewhat organized Git history: all merge commits would correspond only to PRs and we never have merge commits from trunk into feature branches, which *do* muddy the git log. It'll also be easier to find the corresponding PR discussion for any commit in master.

tl;dr:

1. Encourage/enforce smaller PRs to prevent us from getting into a scenario where rebasing is necessary. 

2. Use merge method **only** for final integration of a feature branch into trunk. Visual aid for finding PR discussions from the gitlog.

3. Use rebasing **only** when you need something from trunk in your feature branch and not for feature branch integration. If point 1 is followed/encouraged/enforced, we don't have to get lost in fixing up botched rebases.

What do you think?

**Comment by jcohenadad** [comment]
i really like the idea of merging PRs, so +1 for me. I have often used the feature of searching for PR numbers in old commits to find out when a bug was introduced.

Can we also agree on a "squash and merge", as we already do in SCT and ivadomed? (using the same habits for lab members is also less prone to confusions)

**Comment by jcohenadad** [comment]
> Can we also agree on a "squash and merge", as we already do in SCT and ivadomed? (using the same habits for lab members is also less prone to confusions)

well, i guess https://github.com/shimming-toolbox/shimming-toolbox-py/pull/106#pullrequestreview-498641300 means "no" 😅  ... although i see more pros than cons in squashing (eg: not accumulating fat and unclear commit messages)

**Comment by jidicula** [comment]
> > Can we also agree on a "squash and merge", as we already do in SCT and ivadomed? (using the same habits for lab members is also less prone to confusions)
> 
> well, i guess [#106 (review)](https://github.com/shimming-toolbox/shimming-toolbox-py/pull/106#pullrequestreview-498641300) means "no" 😅 ... although i see more pros than cons in squashing (eg: not accumulating fat and unclear commit messages)

haha yeah, I was just about to mention #106 🙂. If we do want to move towards a squash&merge approach, we should be stricter about coincidental ("I'm in this file so I'll do that extra fix here anyway") fixes. Additionally, we would have to enforce better commit messages; GitHub's squash behaviour is a bit hamfisted in that it just concatenates commit messages into one single commit message. We definitely want to avoid resulting commit messages that would have something like

> Fix function name

> Fix typo

> Remove comment

> Fix function name

where it's unclear which sub-message corresponds to which part of the diff, which gets very messy when using `git blame` locally.

I do like the idea of using squash locally to group "exploratory" commits. So, before converting a PR draft into "ready for review", the programmer does some local squashes to rewrite their local git history, modifies the message for that new squashed commit, then force-pushes their local branch to remote and then requests review.

**Comment by jcohenadad** [comment]
> where it's unclear which sub-message corresponds to which part of the diff, which gets very messy when using git blame locally.

without squashing, those ugly commits would appear in the history, which to me sounds even worst, no?

> I do like the idea of using squash locally to group "exploratory" commits. So, before converting a PR draft into "ready for review", the programmer does some local squashes to rewrite their local git history, modifies the message for that new squashed commit, then force-pushes their local branch to remote and then requests review.

i agree that would be ideal. We've been aiming for that for the past few years, but it just doesn't work, and people keep these commits. I'm the first to blame, I also do that. Why? because it would take me 2 minutes to do a clean squashing, and unfortunately I never have 2 minutes available. So that's where my "let's be pragmatic" tone comes from. So far, I've been pretty happy with the squashing, and I've never noticed any issue with it. Again, I understand the philosophy and push for good practice, but I simply don't have the bandwidth to go over everyone's shoulders to make sure those practices are respected (and I cannot pass on the responsibility to other lab members or software developers, because they already have enough on their plate).

**Comment by jidicula** [comment]
> So that's where my "let's be pragmatic" tone comes from. So far, I've been pretty happy with the squashing, and I've never noticed any issue with it. Again, I understand the philosophy and push for good practice, but I simply don't have the bandwidth to go over everyone's shoulders to make sure those practices are respected (and I cannot pass on the responsibility to other lab members or software developers, because they already have enough on their plate).

Yeah, that's a really good point. Also, unfortunately I haven't come across any really nice Git GUI clients that make these operations more intuitive, except for maybe magit in Emacs, but asking everyone to switch to Emacs just for that would be a bridge too far.

**Comment by po09i** [comment]
To recap, we aim for small PRs and use squash and merge with master.

**Comment by po09i** [comment]
Any cons to removing the ability to rebase and merge in the settings of the repo?

**Comment by jcohenadad** [comment]
> Any cons to removing the ability to rebase and merge in the settings of the repo?

done!
